### ISSUE TITLE
Prevent user from deleting tickets

### ISSUE THREAD

**Issue by wixiwill** [opensupports/opensupports#549]
Hi all,

Is there any way to prevent user from **deleting tickets** ??? 
In my case, it's a big problem if user delets rosolved/closed tickets ... Because it is useful for accounting.

Thanks

**Comment by vish45** [comment]
i was planning to remove the button from the users view but i didnt find it

**Comment by nickbe** [comment]
Actually administrators AND authors MUST be able to delete a ticket. In several countries you HAVE to provide an option for an author to remove his own written words.

**Comment by mredigonda** [comment]
I haven't read all of the GDPR, but I'm wondering if instead of removing the entire ticket we could remove all personal information attached to the ticket (like username, emails, etc.), would it be against the rules?

I understand why @wixiwill and many others would want to keep tickets, not only for accounting purposes, but also to have them as a reference for new problems (maybe new tickets are from people that want to solve a very similar problem to one present in old tickets).

**Comment by versavius** [comment]
> Actually administrators AND authors MUST be able to delete a ticket. In several countries you HAVE to provide an option for an author to remove his own written words.

Yes i agree but in this case this is an audit system so GDPR would not apply, as the recorded comments/posts are evidence to support what was requested.

In systems like this an audit trail is needed, if the author deleted their comments, say they asked for X to be done and they say Y was done, they could delete their own comments showing they actually asked for Y to be and then you have a massive issue or even court case happen and yes i have seen this happen before from users deleting their data and then denying what they said.

For a normal forum yes deleting data is fine but for a ticket thread no.

**Comment by versavius** [comment]
> I haven't read all of the GDPR, but I'm wondering if instead of removing the entire ticket we could remove all personal information attached to the ticket (like username, emails, etc.), would it be against the rules?
> 
> I understand why @wixiwill and many others would want to keep tickets, not only for accounting purposes, but also to have them as a reference for new problems (maybe new tickets are from people that want to solve a very similar problem to one present in old tickets).

As long as you log the relevant details then removing details from the ticket should be ok for example if the user logging the ticket has an account or has one auto generated then their email account is associated to that account so no need to display it in the ticket.

Also if you log the ip address each time a ticket or a ticket comment is raised which is normal for auditing and spam/abuse then no need to have it in the ticket.

But ultimately you can't stop users putting stuff into the ticket comments, thats up to the users to manage not for you to police and manage.

**Comment by nickbe** [comment]
Especially for testing / debuggin / other internal use we must be able to delete tickets for good and without a trace.
But I suggest we simply let the administrator decide if ticket deletion is allowed or not "per department".
That way we can have the bost from both worlds.

**Comment by versavius** [comment]
I think perhaps having a permission to be set per staff member that way you can decide which staff members can delete tickets for admin/testing etc and which can not

**Comment by nickbe** [comment]
No this wouldn't make sense in the current overall concept. Right now you can delete a tickets if you're an administrator. At least this is how it's supposed to be. I would not change this at all.

If an administrator is not able to decide this kind of stuff then he's should not be administrator.  All the audit reasons are useless in my opinion. If I have access to the database I can do whatever I want anyways. So the only thing is to prevent "normal" users from deleting ticket - which should already work as it is.

**Comment by nickbe** [comment]
> I haven't read all of the GDPR, but I'm wondering if instead of removing the entire ticket we could remove all personal information attached to the ticket (like username, emails, etc.), would it be against the rules?
> 
> I understand why @wixiwill and many others would want to keep tickets, not only for accounting purposes, but also to have them as a reference for new problems (maybe new tickets are from people that want to solve a very similar problem to one present in old tickets).

The GDPR would require that we are able to delete an account together with all comments leading to conclusions about the authors identity. Since it's not possible to censor parts of the comments I would recomment that we assign all authors data to a speudo "unknown user" after deleting a user OR a staff member. Yes - staff members are also users with private identities - so their data must be deletable also.
Hence I would assign all sstaff identity to admin when deleting a staff user.

**Comment by versavius** [comment]
why not change the user details to say "unknown user 1", "unknown user 2" etc so you keep all the history but you are renaming the username, first/last name and email address which would remove the user from the system.

Obviously whatever is in a ticket is a different matter but that would be needed for auditing.

**Comment by nickbe** [comment]
Doesn't make any sense. You wouldn't know the identity anyways. 
Auditing is something I stumble upon quite often these days, but honestly this is a simple ticket system.
Was there a problem - was it fixed - maybe how? Quick and easy.

If you're in a company which dies a thorough "auditing" of all its processes and structures this will surely not be the system for you.

**Comment by versavius** [comment]
> Doesn't make any sense. You wouldn't know the identity anyways.
> Auditing is something I stumble upon quite often these days, but honestly this is a simple ticket system.
> Was there a problem - was it fixed - maybe how? Quick and easy.
> 
> If you're in a company which dies a thorough "auditing" of all its processes and structures this will surely not be the system for you.

Agreed, i just thought if we start going down the route of meeting all the GDPR rules etc then where do we stop? and the dev's might be spending more time on coding to meet the rules than having fun on this project.

If a user really wanted to be deleted they could just send an email to the person running their version of opensupports and then the admin could manually rename their account to "unknown user 1" etc surely that would do the same as spending a great deal of time coding a solution?

**Comment by nickbe** [comment]
Exactly. I'm with you a 100 percent here...

**Comment by pieter-groeneweg** [comment]
Maybe it feels like I am breaking in.... but here in NL we just spent a century on getting things right... so a few of my two cents... 

> If a user really wanted to be deleted they could just send an email to the person running their version of opensupports and then the admin could manually rename their account to "unknown user 1" etc surely that would do the same as spending a great deal of time coding a solution?

GDPR rule: have a privacy statement. In this statement:
Tell what you collect on personal data from reporter and the reason why.
Be able to show the collected personal data.
Be able to delete personal data.
Statement: information shared in the tickets will not be used for other means than the issue to be solved. No private data will be shared or stored elsewhere or otherwise unless agreed upon by reporter or explicitly necessary for the resolvement of the incident.

By default, this project only collects an email address and a username.
You can add custom fields of course. As long as you specify what you collect... fine.

But evidently tickets should not be allowed to be deleted by reporter and/or support other than administrators. This to avoid disputes on what was actually requested. Let alone you should be able to justify your actions (to be able to send your invoices ;) ) 

So I'd say... deletion is an admins job, not reporter, reporter can ask you to show what you keep (but that is what they can see when they log in) and ask you to delete his/her personal data... out of business interest you are actually allowed to keep email address and username. (unless you serve private persons)

Add an editable link to show in the bottom black bar to your privacy statement... that solves it all... legally that is.......... that is all there is to GDPR.
